src/mono/browser/runtime/jiterpreter-trace-generator.ts

```diff
@@ -3837,102 +3837,107 @@ function emit_simd_3 (builder: WasmBuilder, ip: MintOpcodePtr, index: SimdIntrin
+                let elementIndex = 0;
+                for (let j = 0; j < elementSize; j++) {
+                    if (j == 0)
+                        elementIndex = nativeIndices[k + j];
+
+                    if (elementIndex >= elementCount || (j > 0 && nativeIndices[k + j] != 0)) {
+                        // this an invalid index make invalid in a safe way
+                        elementIndex = elementCount;
+                        break;
+                    }
+                }
+                for (let j = 0; j < elementSize; j++) {
+                    newShuffleVector[k + j] = (elementIndex * elementSize) + j;
+                }
+            }
+            nativeIndices = newShuffleVector;
         }
         // console.log(`shuffle w/element size ${elementSize} with constant indices ${nativeIndices} (${constantIndices}) -> byte indices ${newShuffleVector}`);
         builder.appendSimd(WasmSimdOpcode.v128_const);
-        builder.appendBytes(newShuffleVector);
+        builder.appendBytes(nativeIndices);
     } else {
         // Load indices (in chars)
         append_ldloc(builder, indicesOffset, WasmOpcode.PREFIX_simd, WasmSimdOpcode.v128_load);
-        // There's no direct narrowing opcode for i32 -> i8, so we have to do two steps :(
-        if (elementCount === 4) {
-            // i32{lane0 ... lane3} -> i16{lane0 ... lane3, 0 ...}
-            builder.v128_const(0);
-            builder.appendSimd(WasmSimdOpcode.i16x8_narrow_i32x4_u);
-        }
-        // Load a zero vector (narrow takes two vectors)
-        builder.v128_const(0);
-        // i16{lane0 ... lane7} -> i8{lane0 ... lane7, 0 ...}
-        builder.appendSimd(WasmSimdOpcode.i8x16_narrow_i16x8_u);
-        // i8{0, 1, 2, 3 ...} -> i8{0, 0, 1, 1, 2, 2, 3, 3 ...}
-        builder.appendSimd(WasmSimdOpcode.v128_const);
-        for (let i = 0; i < elementCount; i++) {
-            for (let j = 0; j < elementSize; j++)
-                builder.appendU8(i);
-        }
-        builder.appendSimd(WasmSimdOpcode.i8x16_swizzle);
-        // multiply indices by 2 or 4 to scale from elt indices to byte indices
-        builder.i32_const(elementCount === 4 ? 2 : 1);
-        builder.appendSimd(WasmSimdOpcode.i8x16_shl);
-        // now add an offset to the additional bytes of each lane, i.e.
-        // 0 1 2 3 0 1 2 3 ...
-        builder.appendSimd(WasmSimdOpcode.v128_const);
-        for (let i = 0; i < elementCount; i++) {
-            for (let j = 0; j < elementSize; j++)
-                builder.appendU8(j);
+        if (elementCount !== 16) {
+            const shift = elementCount === 8 ? 1 : elementCount === 4 ? 2 : 3;
+
+            if (checkBounds) {
+                builder.local("shuffle_indices", WasmOpcode.tee_local);
+
+                builder.appendSimd(WasmSimdOpcode.v128_const);
+                for (let i = 0; i < elementCount; i++) {
+                    builder.appendU8(elementCount - 1);
```

Comment:
I recommend generating an integer constant of the appropriate size and splatting it to generate the vector constant